dan_layer/core/src/workers/pacemaker_worker.rs

```diff
@@ -0,0 +1,225 @@
+// Copyright 2021. The Tari Project
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use std::{collections::HashSet, time::Duration};
+
+use log::*;
+use tari_shutdown::ShutdownSignal;
+use tokio::{
+    sync::mpsc::{Receiver, Sender},
+    task::JoinHandle,
+};
+
+use super::hotstuff_error::HotStuffError;
+use crate::models::pacemaker::{PacemakerSignal, PacemakerWaitStatus, WaitOver};
+
+const LOG_TARGET: &str = "tari::dan_layer::pacemaker_worker";
+
+#[derive(Debug)]
+pub struct LeaderFailurePacemaker {
+    rx_waiter_signal: Receiver<PacemakerSignal>,
+    tx_waiter_status: Sender<(WaitOver, PacemakerWaitStatus)>,
+    max_timeout: u64,
+    wait_over_set: HashSet<WaitOver>,
+}
+
+impl LeaderFailurePacemaker {
```

Comment:
Lets call it just Pacemaker. Btw you never actually know that any leader failed.

Reply:
Cool, let's rename it. Yes, we just timeout on the receiving of `StartWait` messages.